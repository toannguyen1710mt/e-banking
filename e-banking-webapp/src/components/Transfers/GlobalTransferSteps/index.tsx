'use client';

// Libs
import { useForm } from 'react-hook-form';
import { z } from 'zod';
import { zodResolver } from '@hookform/resolvers/zod';
import { Session } from 'next-auth';

// Constants
import { ERROR_MESSAGES } from '@/constants';

// Models
import { IAccountPayloadData, TransactionCreateData } from '@/interfaces';

// API
import { createTransaction, updateAccountInfo } from '@/actions';

// Helpers / Utils
import { GlobalTransferFormSchema } from '@/schemas';
import {
  convertToUSD,
  formatNumberWithCommas,
  sanitizeAmount,
  toastManager,
} from '@/utils';

// Components
import * as WizardForm from '@/components/common/WizardForm';
import {
  GlobalTransferForm,
  ConfirmGlobalTransfer,
  GlobalTransferSuccess,
} from '@/components';

// Context
import { FetchedBalancesProvider } from '@/context';

type FormValues = z.infer<typeof GlobalTransferFormSchema>;

interface IGlobalTransferSteps {
  session: Session;
  onClose?: () => void;
}

export const GlobalTransferSteps = ({
  session,
  onClose,
}: IGlobalTransferSteps) => {
  const form = useForm<FormValues>({
    mode: 'onBlur',
    defaultValues: {
      globalTransfer: {
        fromAccountType: undefined,
        fromCountryType: undefined,
        amount: '',
      },
    },
    resolver: zodResolver(GlobalTransferFormSchema),
  });

  const allFieldValues = form.watch();

  const amountInUSD = formatNumberWithCommas(
    convertToUSD(
      allFieldValues.globalTransfer.fromCountryType,
      Number(allFieldValues.globalTransfer.amount),
    ),
  );

  const sanitizedAmountInUSD = sanitizeAmount(amountInUSD);

  const submitHandler = async (data: FormValues) => {
    const {
      fromAccountId,
      fromCardName,
      fromAccountNumber,
      fromAccountBalance,
      globalTransfer: { fromAccountType, amount },
    } = data;

    const transactionData: TransactionCreateData = {
      fromAccountId,
      toAccountId: allFieldValues.globalTransfer.recipientAccount,
      fromAccountType,
      toAccountType: undefined,
      statusTransaction: true,
      amount: sanitizeAmount(
        formatNumberWithCommas(
          convertToUSD(
            allFieldValues.globalTransfer.fromCountryType,
            Number(amount),
          ),
        ),
      ),
      recipientName: allFieldValues.recipientName,
    };

    const payload: IAccountPayloadData = {
      name: fromCardName,
      accountNumber: fromAccountNumber,
      balance: fromAccountBalance - sanitizedAmountInUSD,
      type: fromAccountType,
      currency: '$',
    };

    try {
      await createTransaction(fromAccountId, transactionData);
      await updateAccountInfo(fromAccountId, payload);
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (error) {
      toastManager.showToast(ERROR_MESSAGES.TRANSFER_FAILED);
    }
  };

  return (
    <FetchedBalancesProvider>
      <WizardForm.Root
        schema={GlobalTransferFormSchema}
        form={form}
        className='flex grow flex-col'
      >
        <WizardForm.Step name='transfer' key='transfer'>
          <GlobalTransferForm session={session} />
        </WizardForm.Step>
        <WizardForm.Step
          name='confirm'
          key='confirm'
          className='flex grow flex-col items-center justify-center'
        >
          <ConfirmGlobalTransfer
            {...allFieldValues}
            submitHandler={submitHandler}
            amountInUSD={amountInUSD}
          />
        </WizardForm.Step>
        <WizardForm.Step
          name='success'
          key='success'
          className='flex grow flex-col items-center justify-center'
        >
          <GlobalTransferSuccess {...allFieldValues} onClose={onClose} />
        </WizardForm.Step>
      </WizardForm.Root>
    </FetchedBalancesProvider>
  );
};
