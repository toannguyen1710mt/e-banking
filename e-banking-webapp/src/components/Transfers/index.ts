export * from './InternalTransferForm';
export * from './ConfirmInternalTransfer';
export * from './ConfirmGlobalTransfer';
export * from './InternalTransferSuccess';
export * from './GlobalTransferForm';
export * from './GlobalTransferSuccess';
export * from './TransferModal';
export * from './TransferTab';
export * from './GlobalTransferSteps';
export * from './InternalTransferSteps';
export * from './TransferTable';
