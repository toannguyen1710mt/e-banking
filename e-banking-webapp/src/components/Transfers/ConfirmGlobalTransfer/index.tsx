'use client';

// Third Party
import { z } from 'zod';
import { useTransition } from 'react';

// Utils
import { GlobalTransferFormSchema } from '@/schemas';

// Hooks
import { useWizardFormContext } from '@/context';

// Components
import { Button, Text } from '@/components';

interface ConfirmGlobalTransferProps<T extends z.ZodType> {
  submitHandler: (data: z.infer<T>) => void;
  amountInUSD: string;
}

export const ConfirmGlobalTransfer = <T extends z.ZodType>({
  submitHandler,
  amountInUSD,
}: ConfirmGlobalTransferProps<T>) => {
  const {
    form: { getValues },
    onPrevStep,
    onNextStep,
  } = useWizardFormContext<typeof GlobalTransferFormSchema>();

  const [isPending, startTransition] = useTransition();

  const values = getValues();

  const {
    globalTransfer: { fromAccountType },
    recipientName,
  } = values || {};

  const onSubmit = (e: React.FormEvent) => {
    e.preventDefault();

    startTransition(() => {
      const data = getValues();
      submitHandler(data);
      onNextStep();
    });
  };

  return (
    <div className='flex w-[252px] flex-col items-center text-center'>
      <Text className='text-sm font-semibold text-navyBlue'>
        {recipientName} is about to receive
      </Text>
      <Text className='mt-3 text-4xl font-bold text-navyBlue'>
        ${amountInUSD}
      </Text>
      <Text className='mt-3 text-xs font-medium text-transparentBlack'>
        From your {fromAccountType} wallet to {recipientName}&apos;s wallet,
        this action cannot be undone once approved...
      </Text>
      <div className='mt-10 flex gap-6'>
        <Button
          radius='xs'
          color='tertiary'
          size='md'
          isDisabled={isPending}
          onClick={onPrevStep}
        >
          Cancel
        </Button>
        <Button
          radius='xs'
          color='navyBlue'
          size='md'
          onClick={onSubmit}
          isLoading={isPending}
        >
          Proceed
        </Button>
      </div>
    </div>
  );
};
