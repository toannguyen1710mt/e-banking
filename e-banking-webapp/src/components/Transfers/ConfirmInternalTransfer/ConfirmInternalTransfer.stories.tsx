// Libs
import type { Meta, StoryObj } from '@storybook/react';

// Utils
import { InternalTransferFormSchema } from '@/schemas';

// Mocks
import { useMockFormInternal } from '@/mocks';

// Components
import { ConfirmInternalTransfer } from '@/components';
import * as WizardForm from '@/components/common/WizardForm';

const meta = {
  title: 'Components/Transfers/ConfirmInternalTransfer',
  tags: ['autodocs'],
  component: ConfirmInternalTransfer,
  parameters: {
    layout: 'centered',
  },
} satisfies Meta<typeof ConfirmInternalTransfer>;

export default meta;

type Story = StoryObj<typeof ConfirmInternalTransfer>;

export const Default: Story = {
  render: (args) => (
    <WizardForm.Root
      schema={InternalTransferFormSchema}
      form={useMockFormInternal()}
      className='flex grow flex-col'
    >
      <WizardForm.Step
        name='confirm'
        key='confirm'
        className='flex grow flex-col items-center justify-center'
      >
        <ConfirmInternalTransfer {...args} />
      </WizardForm.Step>
    </WizardForm.Root>
  ),
};
