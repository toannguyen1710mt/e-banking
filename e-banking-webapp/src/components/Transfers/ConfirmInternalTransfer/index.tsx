'use client';

// Libs
import { useTransition } from 'react';
import { z } from 'zod';

// Components
import { Button, Text } from '@/components';

// Utils
import { formatNumberWithCommas } from '@/utils';

// Contexts
import { useWizardFormContext } from '@/context';

// Schemas
import { InternalTransferFormSchema } from '@/schemas';

interface IConfirmInternalTransferProps<T extends z.ZodType> {
  submitHandler: (data: z.infer<T>) => void;
}

export const ConfirmInternalTransfer = <T extends z.ZodType>({
  submitHandler,
}: IConfirmInternalTransferProps<T>) => {
  const {
    form: { getValues },
    onPrevStep,
    onNextStep,
  } = useWizardFormContext<typeof InternalTransferFormSchema>();

  const [isPending, startTransition] = useTransition();

  const values = getValues();

  const onSubmit = (e: React.FormEvent) => {
    e.preventDefault();

    startTransition(() => {
      const data = getValues();
      submitHandler(data);
      onNextStep();
    });
  };

  return (
    <div className='flex flex-col items-center justify-center gap-4'>
      <Text as='h4' className='text-sm font-semibold'>
        You are about to transfer
      </Text>
      <Text as='span' className='text-4xl font-bold'>
        ${formatNumberWithCommas(Number(values.internalTransfer.amount))}
      </Text>
      <Text className='max-w-[250px] text-center text-xs font-medium text-primary-200 opacity-50'>
        From your {values.internalTransfer.fromAccountType} wallet to your&nbsp;
        {values.internalTransfer.toAccountType} wallet, this action cannot be
        undone once approved...
      </Text>
      <div className='mt-8 flex gap-8'>
        <Button
          color='tertiary'
          radius='xs'
          size='md'
          isDisabled={isPending}
          onClick={onPrevStep}
        >
          Cancel
        </Button>
        <Button
          aria-label='proceed'
          color='primary'
          radius='xs'
          size='md'
          isDisabled={isPending}
          isLoading={isPending}
          onClick={onSubmit}
        >
          Proceed
        </Button>
      </div>
    </div>
  );
};
