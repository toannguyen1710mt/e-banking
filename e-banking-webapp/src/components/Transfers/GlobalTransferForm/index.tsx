'use client';

// Third Party
import { Controller, useWatch } from 'react-hook-form';
import { Session } from 'next-auth';
import { z } from 'zod';
import { useEffect, useState, useTransition } from 'react';
import { Spinner } from '@nextui-org/react';

// Constants
import {
  ERROR_MESSAGES,
  OPTIONS_COUNTRY_CODE_CONVERT_GLOBAL,
  TRANSFER_FORM_ACCOUNT_OPTIONS,
  TRANSFER_FORM_GLOBAL_OPTIONS,
} from '@/constants';

// Interfaces
import { AccountType, GlobalAccount, GlobalType } from '@/interfaces';

// API
import { getAccountInfoByAccountType, getGlobalAccounts } from '@/services';

// Helpers / Utils
import {
  convertToUSD,
  formatNumberWithCommas,
  isValidNumber,
  sanitizeNumber,
} from '@/utils';
import { GlobalTransferFormSchema } from '@/schemas';

// Components
import { Button, Input, Select, Text, SendIcon } from '@/components';

// Contexts
import { useFetchedBalances, useWizardFormContext } from '@/context';

type FormValues = keyof z.infer<typeof GlobalTransferFormSchema>;

export const GlobalTransferForm = ({ session }: { session: Session }) => {
  const {
    form: {
      control,
      formState: { errors },
      setValue,
      getValues,
      setError,
      clearErrors,
    },
    onNextStep,
  } = useWizardFormContext<typeof GlobalTransferFormSchema>();

  const hiddenFields: FormValues[] = [
    'fromAccountId',
    'fromCardName',
    'fromAccountNumber',
    'fromAccountBalance',
    'recipientName',
  ];

  const fromAccountTypeValue = useWatch({
    control,
    name: 'globalTransfer.fromAccountType',
  });

  const fromCountryType = useWatch({
    control,
    name: 'globalTransfer.fromCountryType',
  });

  const allFieldValues = getValues();

  const amountValue = allFieldValues.globalTransfer.amount;

  const amountValueInUSD = convertToUSD(
    allFieldValues.globalTransfer.fromCountryType,
    Number(allFieldValues.globalTransfer.amount),
  );

  const [isPending, startTransition] = useTransition();

  // States for global accounts
  const [isFetchingGlobalAccounts, setIsFetchingGlobalAccounts] =
    useState(false);
  const [selectedGlobalAccount, setSelectedGlobalAccount] =
    useState<GlobalAccount | null>(null);

  // States for balances
  const [balanceSend, setBalanceSend] = useState<number | null>(null);
  const [rawAmount, setRawAmount] = useState<string>('');
  const [amountError, setAmountError] = useState<string | null>(null);

  // Cached balance data
  const { fetchedBalances, setFetchedBalances } = useFetchedBalances();

  useEffect(() => {
    const fetchBalanceSend = async () => {
      if (
        !fromAccountTypeValue ||
        fetchedBalances[fromAccountTypeValue] !== undefined
      ) {
        setBalanceSend(fetchedBalances[fromAccountTypeValue] || null);
        return;
      }

      startTransition(async () => {
        try {
          const balance = await getAccountInfoByAccountType(
            session.user.id,
            fromAccountTypeValue,
            'balance',
          );

          const accountId = await getAccountInfoByAccountType(
            session.user.id,
            fromAccountTypeValue,
            'documentId',
          );

          const fromCardName = await getAccountInfoByAccountType(
            session.user.id,
            fromAccountTypeValue,
            'name',
          );

          const fromAccountNumber = await getAccountInfoByAccountType(
            session.user.id,
            fromAccountTypeValue,
            'accountNumber',
          );

          startTransition(() => {
            setBalanceSend(Number(balance));
            setFetchedBalances((prev) => ({
              ...prev,
              [fromAccountTypeValue]: Number(balance),
            }));
            setValue('fromAccountId', String(accountId));
            setValue('fromCardName', String(fromCardName));
            setValue('fromAccountNumber', String(fromAccountNumber));
            setValue('fromAccountBalance', Number(balance));
          });
        } catch (error) {
          console.error(ERROR_MESSAGES.GET_BALANCE_FOR_ACCOUNT, error);
        }
      });
    };

    fetchBalanceSend();
  }, [
    fromAccountTypeValue,
    session.user.id,
    setValue,
    fetchedBalances,
    setFetchedBalances,
  ]);

  const countryCode = () => {
    return (
      OPTIONS_COUNTRY_CODE_CONVERT_GLOBAL.find(
        (option) => option.key === fromCountryType,
      )?.label || ''
    );
  };

  const filteredFromAccountOptions = () =>
    TRANSFER_FORM_ACCOUNT_OPTIONS.filter(
      (option) => option.key !== (fromCountryType as unknown as AccountType),
    );

  const filteredToAccountOptions = () =>
    TRANSFER_FORM_GLOBAL_OPTIONS.filter(
      (option) =>
        option.key !== (fromAccountTypeValue as unknown as GlobalType),
    );

  const handleAmountErrors = (balanceSend: number | null, amount: number) => {
    setAmountError(
      balanceSend && amount > balanceSend
        ? `${ERROR_MESSAGES.AMOUNT_EXCEEDED_BALANCE} $${formatNumberWithCommas(balanceSend)}`
        : null,
    );
  };

  const handleInputChange = (
    value: string,
    onChange: (value: string) => void,
  ) => {
    const sanitizedValue = sanitizeNumber(value);

    if (isValidNumber(sanitizedValue)) {
      setRawAmount(sanitizedValue);

      handleAmountErrors(balanceSend, amountValueInUSD);

      onChange(sanitizedValue);
    }
  };

  const getFormattedAmount = (
    rawAmount: string,
    value: string | number,
  ): string => {
    if (rawAmount) {
      return formatNumberWithCommas(Number(rawAmount));
    }

    if (String(value) !== '') {
      return formatNumberWithCommas(Number(value));
    }

    return '';
  };

  const handleAccountErrors = (accountMatch: GlobalAccount | undefined) => {
    if (accountMatch) {
      setSelectedGlobalAccount(accountMatch);
      clearErrors('globalTransfer.recipientAccount');
    } else {
      setError('globalTransfer.recipientAccount', {
        message: ERROR_MESSAGES.RECIPIENT_ACCOUNT_INVALID,
      });
    }
  };

  const validateRecipientAccount = async (value: string) => {
    if (!value) {
      setSelectedGlobalAccount(null);
      setError('globalTransfer.recipientAccount', {
        message: ERROR_MESSAGES.FIELD_REQUIRED,
      });
      return;
    }

    setIsFetchingGlobalAccounts(true);
    try {
      const response = await getGlobalAccounts();

      const accountMatch = response.data.find(
        (account) =>
          account.accountNumber === value && account.currency === countryCode(),
      );

      handleAccountErrors(accountMatch);
    } catch (error) {
      console.error('Failed to fetch global accounts:', error);
    } finally {
      setIsFetchingGlobalAccounts(false);
    }
  };

  useEffect(() => {
    if (fromAccountTypeValue && amountValue) {
      handleAmountErrors(balanceSend, amountValueInUSD);
    }
  }, [amountValueInUSD, balanceSend, fromAccountTypeValue, amountValue]);

  useEffect(() => {
    if (fromCountryType && getValues('globalTransfer.recipientAccount')) {
      validateRecipientAccount(getValues('globalTransfer.recipientAccount'));
    }
  }, [fromCountryType]);

  useEffect(() => {
    if (selectedGlobalAccount) {
      setValue('recipientName', selectedGlobalAccount.name);
    }
  }, [selectedGlobalAccount, setValue]);

  const disableButtonSubmit =
    !fromAccountTypeValue ||
    !fromCountryType ||
    !allFieldValues.globalTransfer.recipientAccount ||
    !allFieldValues.globalTransfer.amount ||
    !!errors.globalTransfer?.recipientAccount ||
    !!errors.globalTransfer?.amount ||
    !!amountError;

  return (
    <div className='flex w-full flex-col gap-4'>
      {/* Title */}
      <Text as='h4' className='text-xs font-medium'>
        Send Money Across The Global with Ease
      </Text>

      {/* Country */}
      <Controller
        control={control}
        name='globalTransfer.fromCountryType'
        render={({ field: { onChange, onBlur, value } }) => {
          return (
            <Select
              label='Country'
              labelPlacement='outside'
              options={filteredToAccountOptions()}
              classNames={{ label: 'text-sm' }}
              value={String(value)}
              errorMessage={errors.globalTransfer?.fromCountryType?.message}
              isInvalid={!!errors.globalTransfer?.fromCountryType}
              selectedKeys={value ? [String(value)] : []}
              onSelectionChange={(keys) => {
                const selectedValue = String(Array.from(keys)[0]);
                onChange(selectedValue);
              }}
              onClose={onBlur}
            />
          );
        }}
      />

      {/* Account */}
      <Controller
        control={control}
        name='globalTransfer.fromAccountType'
        render={({ field: { onChange, onBlur, value } }) => {
          return (
            <Select
              label='Account'
              labelPlacement='outside'
              options={filteredFromAccountOptions()}
              classNames={{ label: 'text-sm' }}
              value={String(value)}
              errorMessage={errors.globalTransfer?.fromAccountType?.message}
              isInvalid={!!errors.globalTransfer?.fromAccountType}
              selectedKeys={value ? [String(value)] : []}
              onSelectionChange={(keys) => {
                const selectedValue = String(Array.from(keys)[0]);
                onChange(selectedValue);
              }}
              onClose={onBlur}
            />
          );
        }}
      />

      {/* Recipient Account */}
      {fromAccountTypeValue && (
        <Controller
          control={control}
          name='globalTransfer.recipientAccount'
          render={({ field: { value, onChange } }) => {
            return (
              <Input
                label='Recipient Account'
                labelPlacement='outside'
                placeholder=' '
                classNames={{
                  inputWrapper:
                    'h-10 px-2.5 py-2 rounded-sm border-default box-border',
                  input: 'm-0 text-xs text-primary-200 font-medium',
                }}
                value={value}
                errorMessage={errors.globalTransfer?.recipientAccount?.message}
                isInvalid={!!errors.globalTransfer?.recipientAccount}
                onChange={onChange}
                onBlur={() => {
                  validateRecipientAccount(value);
                }}
                maxLength={12}
              />
            );
          }}
        />
      )}

      {/* Available Balance */}
      <div className='flex items-center gap-2'>
        <Text as='span' className='text-xs text-foreground-200 opacity-50'>
          Available Balance:
        </Text>
        <Text as='span' className='text-xs text-foreground-200 opacity-50'>
          {isPending ? (
            <Spinner size='sm' />
          ) : (
            balanceSend && `$${formatNumberWithCommas(balanceSend)}`
          )}
        </Text>
      </div>

      {/* Amount */}
      <Controller
        control={control}
        name='globalTransfer.amount'
        render={({
          field: { onBlur, value, onChange },
          fieldState: { error },
        }) => {
          return (
            <div className='flex items-baseline gap-[15px]'>
              <Input
                value={countryCode()}
                readOnly
                classNames={{
                  base: 'w-11',
                  inputWrapper:
                    'w-11 h-10 rounded-sm border-default box-border p-0',
                  input: 'font-semibold text-navyBlue text-xs text-center',
                }}
              />
              <Input
                inputMode='decimal'
                label='Amount'
                labelPlacement='outside'
                placeholder=' '
                classNames={{
                  inputWrapper:
                    'h-10 px-2.5 py-2 rounded-sm border-default box-border',
                  input: 'm-0 text-xs text-primary-200 font-medium',
                }}
                value={getFormattedAmount(rawAmount, value)}
                onChange={(e) =>
                  handleInputChange(e.target.value.replace(/^\$/, ''), onChange)
                }
                errorMessage={amountError || error?.message}
                isInvalid={!!amountError || !!error?.message}
                onBlur={onBlur}
              />
            </div>
          );
        }}
      />

      {/* Hidden Fields */}
      {hiddenFields.map((fieldName) => (
        <Controller
          key={fieldName}
          control={control}
          name={fieldName}
          render={({ field: { value } }) => (
            <Input
              value={
                fieldName === 'recipientName'
                  ? getValues('recipientName')
                  : String(value)
              }
              className='hidden'
            />
          )}
        />
      ))}

      <Button
        type='submit'
        startContent={<SendIcon />}
        className='bg-primary-200 font-semibold text-foreground-200'
        onClick={onNextStep}
        isDisabled={disableButtonSubmit}
      >
        Transfer Funds
      </Button>

      {isFetchingGlobalAccounts && (
        <div className='absolute inset-0 z-40 flex items-center justify-center rounded-xl bg-background-400/30'>
          <Spinner size='sm' color='success' />
        </div>
      )}
    </div>
  );
};
