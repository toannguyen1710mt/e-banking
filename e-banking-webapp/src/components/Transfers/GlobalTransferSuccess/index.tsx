'use client';

// Components
import { Button, CheckIcon, Text } from '@/components';

// Contexts
import { useWizardFormContext } from '@/context';

// Schemas
import { GlobalTransferFormSchema } from '@/schemas';

interface IGlobalTransferSuccess {
  onClose?: () => void;
}

export const GlobalTransferSuccess = ({ onClose }: IGlobalTransferSuccess) => {
  const { form } = useWizardFormContext<typeof GlobalTransferFormSchema>();
  const values = form.getValues();

  const {
    globalTransfer: { amount },
    recipientName,
  } = values;

  return (
    <div className='flex flex-col items-center gap-4'>
      <div className='flex h-14 w-14 items-center justify-center rounded-full bg-background-300'>
        <CheckIcon />
      </div>
      <Text as='h4' className='text-sm font-semibold'>
        Transaction was Successful
      </Text>
      <Text className='max-w-[220px] text-center text-xs font-medium text-primary-200 opacity-50'>
        A sum of ${amount} was transferred to {recipientName}
      </Text>

      <Button
        color='tertiary'
        radius='xs'
        size='md'
        className='mt-8'
        onClick={onClose}
      >
        Close
      </Button>
    </div>
  );
};
