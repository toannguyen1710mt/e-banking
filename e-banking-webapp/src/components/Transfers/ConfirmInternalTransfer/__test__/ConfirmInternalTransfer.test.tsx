// Libs
import { fireEvent, render, screen } from '@testing-library/react';

// Interfaces
import { AccountType } from '@/interfaces';

// Components
import { ConfirmInternalTransfer } from '@/components/Transfers/ConfirmInternalTransfer';

// Contexts
import { useWizardFormContext } from '@/context';

// Schemas
import { InternalTransferFormSchema } from '@/schemas';

jest.mock('@/context', () => ({
  useWizardFormContext: jest.fn(),
}));

describe('ConfirmInternalTransfer component', () => {
  const mockFormValues = {
    internalTransfer: {
      fromAccountType: AccountType.MAIN,
      toAccountType: AccountType.SAVINGS,
      amount: '1500',
    },
    fromAccountId: 'acc-123',
    toAccountId: 'acc-456',
    fromCardName: 'Main Account',
    toCardName: 'Savings Account',
    fromAccountNumber: '123456789',
    toAccountNumber: '987654321',
    fromAccountBalance: 2000,
    toAccountBalance: 1000,
  };

  const mockGetValues = jest.fn().mockReturnValue(mockFormValues);
  const mockOnPrevStep = jest.fn();
  const mockOnNextStep = jest.fn();
  const mockSubmitHandler = jest.fn();

  beforeEach(() => {
    (useWizardFormContext as jest.Mock).mockReturnValue({
      form: { getValues: mockGetValues },
      onPrevStep: mockOnPrevStep,
      onNextStep: mockOnNextStep,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const { container } = render(
      <ConfirmInternalTransfer<typeof InternalTransferFormSchema>
        submitHandler={mockSubmitHandler}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should call onPrevStep when clicking Cancel button', () => {
    render(
      <ConfirmInternalTransfer<typeof InternalTransferFormSchema>
        submitHandler={mockSubmitHandler}
      />,
    );

    fireEvent.click(screen.getByText('Cancel'));
    expect(mockOnPrevStep).toHaveBeenCalledTimes(1);
  });

  it('should call submitHandler and onNextStep when Proceed button is clicked', () => {
    render(
      <ConfirmInternalTransfer<typeof InternalTransferFormSchema>
        submitHandler={mockSubmitHandler}
      />,
    );

    const proceedButton = screen.getByText('Proceed');
    fireEvent.click(proceedButton);

    // Verify form submission behavior
    expect(mockGetValues).toHaveBeenCalled();
    expect(mockSubmitHandler).toHaveBeenCalledWith(mockFormValues);
    expect(mockOnNextStep).toHaveBeenCalled();
  });
});
