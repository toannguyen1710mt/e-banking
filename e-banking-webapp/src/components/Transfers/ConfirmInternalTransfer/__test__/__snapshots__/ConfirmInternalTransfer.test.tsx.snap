// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmInternalTransfer component should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-center gap-4"
  >
    <h4
      class="text-primary-200 text-sm font-semibold"
    >
      You are about to transfer
    </h4>
    <span
      class="text-primary-200 text-4xl font-bold"
    >
      $
      1,500
    </span>
    <p
      class="text-primary-200 max-w-[250px] text-center text-xs font-medium text-primary-200 opacity-50"
    >
      From your 
      Main
       wallet to yourÂ 
      Savings
       wallet, this action cannot be undone once approved...
    </p>
    <div
      class="mt-8 flex gap-8"
    >
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 gap-2 [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover border-transparent bg-background-800 text-foreground-200 rounded m-w-[69px] h-[35px] py-2.5 text-xs font-semibold max-w-24 w-auto"
        type="button"
      >
        Cancel
      </button>
      <button
        aria-label="proceed"
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 gap-2 [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover border-transparent bg-background-300 text-foreground-200 rounded m-w-[69px] h-[35px] py-2.5 text-xs font-semibold max-w-24 w-auto"
        type="button"
      >
        Proceed
      </button>
    </div>
  </div>
</div>
`;
